PhysicsTools/PyTorch/test/testModelJit.cc

```diff
@@ -0,0 +1,166 @@
+#include <cuda_runtime.h>
+#include <c10/cuda/CUDAStream.h>
+#include "PhysicsTools/PyTorch/interface/Model.h"
+#include "PhysicsTools/PyTorch/test/Nvtx.h"
+#include "PhysicsTools/PyTorch/test/testTorchBase.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/requireDevices.h"
+
+namespace torchtest {
+
+  using namespace cms::torch;
+
+  class TestModelJIT : public testTorchBase {
+  public:
+    std::string script() const override;
+
+    void testCtor_DefaultDeviceIsCpu();
+    void testCtor_ExplicitDeviceIsHonored();
+    void testCtor_BadModelPathThrows();
+    void testToDevice_UpdatesUnderlyingState();
+    void testToDevice_NonBlocking();
+    void testForward_IdempotentOutput();
+    void testForward_OutputOnCorrectDevice();
+    void testAsyncExecution();
+
+  private:
+    CPPUNIT_TEST_SUITE(TestModelJIT);
+    CPPUNIT_TEST(testCtor_DefaultDeviceIsCpu);
+    CPPUNIT_TEST(testCtor_ExplicitDeviceIsHonored);
+    CPPUNIT_TEST(testCtor_BadModelPathThrows);
+    CPPUNIT_TEST(testToDevice_UpdatesUnderlyingState);
+    CPPUNIT_TEST(testToDevice_NonBlocking);
+    CPPUNIT_TEST(testForward_IdempotentOutput);
+    CPPUNIT_TEST(testForward_OutputOnCorrectDevice);
+    CPPUNIT_TEST(testAsyncExecution);
+    CPPUNIT_TEST_SUITE_END();
+
+    const int64_t batch_size_ = 2 << 10;
+  };
+
+  CPPUNIT_TEST_SUITE_REGISTRATION(TestModelJIT);
+
+  std::string TestModelJIT::script() const { return "testExportLinearDnn.py"; }
+
+  void TestModelJIT::testCtor_DefaultDeviceIsCpu() {
+    auto m_path = modelPath() + "/linear_dnn.pt";
+    auto m = Model(m_path);
+
+    CPPUNIT_ASSERT_EQUAL(::torch::kCPU, m.device().type());
+  }
+
+  void TestModelJIT::testCtor_ExplicitDeviceIsHonored() {
+    // disable test on non-CUDA devices
+    if (!cms::cudatest::testDevices())
+      return;
+
+    auto m_path = modelPath() + "/linear_dnn.pt";
+    auto dev = ::torch::Device(::torch::kCUDA, 0);
+    auto m = Model(m_path, dev);
+
+    CPPUNIT_ASSERT_EQUAL(dev, m.device());
+  }
+
+  void TestModelJIT::testCtor_BadModelPathThrows() {
+    auto m_path = modelPath() + "/not_existing_model.pt";
+    CPPUNIT_ASSERT_THROW(Model m(m_path), cms::Exception);
+  }
+
+  void TestModelJIT::testToDevice_UpdatesUnderlyingState() {
+    // disable test on non-CUDA devices
+    if (!cms::cudatest::testDevices())
+      return;
+
+    auto m_path = modelPath() + "/linear_dnn.pt";
+    auto m = Model(m_path);
+    auto dev = ::torch::Device(::torch::kCUDA, 0);
```

Comment:
If the system has more than one (visible) GPU, would it be feasible to run the tests on all of them ?